File: examples/go/cmd/server/dex/engagement_controller.go
// Copyright (c) 2022-2026 Super Durable, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

package dex

import (
	"github.com/gin-gonic/gin"
	"github.com/superdurable/dex/examples/go/workflows"
	"github.com/superdurable/dex/examples/go/workflows/engagement"
	sdk "github.com/superdurable/dex/sdk-go/dex"
)

type engagementController struct {
	client *sdk.Client
}

func newEngagementController(client *sdk.Client) *engagementController {
	return &engagementController{client: client}
}

func (controller *engagementController) registerRoutes(router *gin.Engine) {
	router.GET("/engagement/start", controller.start)
	router.GET("/engagement/describe", controller.describe)
	router.GET("/engagement/optout", controller.optOutReminder)
	router.GET("/engagement/decline", controller.decline)
	router.GET("/engagement/accept", controller.accept)
	router.GET("/engagement/list", controller.list)
}

func (controller *engagementController) start(request *gin.Context) {
	flowID := newFlowID("engagement")
	input := engagement.EngagementInput{
		EmployerID:  "test-employer-id",
		JobSeekerID: "test-job-seeker-id",
		Notes:       "test-notes",
	}
	startFlow(request, controller.client, workflows.Engagement, flowID, input)
}

func (controller *engagementController) describe(request *gin.Context) {
	flowID, found := requiredQuery(request, "workflowId")
	if !found {
		return
	}
	var output engagement.EngagementDescription
	err := controller.client.InvokeRPC(
		request.Request.Context(),
		flowID,
		workflows.Engagement.Describe,
		nil,
		&output,
		sdk.InvokeOptions{},
	)
	respond(request, output, err)
}

func (controller *engagementController) optOutReminder(request *gin.Context) {
	flowID, found := requiredQuery(request, "workflowId")
	if !found {
		return
	}
	err := controller.client.PublishToChannel(
		request.Request.Context(),
		flowID,
		engagement.OptOutReminder,
		nil,
	)
	respond(request, struct{}{}, err)
}

func (controller *engagementController) decline(request *gin.Context) {
	controller.update(request, workflows.Engagement.Decline)
}

func (controller *engagementController) accept(request *gin.Context) {
	controller.update(request, workflows.Engagement.Accept)
}

func (controller *engagementController) update(
	request *gin.Context,
	rpc sdk.RPC[string, engagement.Status],
) {
	flowID, found := requiredQuery(request, "workflowId")
	if !found {
		return
	}
	var output engagement.Status
	err := controller.client.InvokeRPC(
		request.Request.Context(),
		flowID,
		rpc,
		request.Query("notes"),
		&output,
		sdk.InvokeOptions{},
	)
	respond(request, output, err)
}

func (controller *engagementController) list(request *gin.Context) {
	query, found := requiredQuery(request, "query")
	if !found {
		return
	}
	page, err := controller.client.SearchFlows(request.Request.Context(), query, 100, "")
	respond(request, page, err)
}

File: examples/java/src/main/java/io/superdurable/dex/products/engagement/EngagementController.java
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.superdurable.dex.products.engagement;

import io.superdurable.dex.Client;
import io.superdurable.dex.SearchFlowsPage;
import io.superdurable.dex.shared.ExampleFlows;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.time.Duration;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;

@RestController
@RequestMapping("/products/engagement")
public class EngagementController {
    private final Client client;
    private final EngagementFlow flow;

    public EngagementController(
            final Client client,
            final EngagementFlow flow) {
        this.client = client;
        this.flow = flow;
    }

    @GetMapping("/start")
    public ResponseEntity<Map<String, String>> start() {
        final String flowId = "engagement-" + System.nanoTime();
        final EngagementInput input = new EngagementInput(
                "test-employer-id",
                "test-job-seeker-id",
                "test-notes");
        final String runId = client.startFlow(flow, flowId, input, ExampleFlows.startOptions());
        client.waitForAttributeEqual(
                flowId,
                flow.employerId,
                input.employerId,
                Duration.ofSeconds(15));
        final Map<String, String> response = new LinkedHashMap<String, String>();
        response.put("flowID", flowId);
        response.put("runID", runId);
        return ResponseEntity.ok(response);
    }

    @GetMapping("/describe")
    public ResponseEntity<EngagementDescription> describe(
            @RequestParam final String workflowId) {
        final EngagementFlow stub = client.newRpcStub(EngagementFlow.class, workflowId);
        return ResponseEntity.ok(client.invokeRPC(stub::describe));
    }

    @GetMapping("/optout")
    public ResponseEntity<Map<String, Object>> optOut(
            @RequestParam final String workflowId) {
        client.publish(workflowId, flow.optOutReminder, (Void) null);
        return ResponseEntity.ok(Collections.<String, Object>emptyMap());
    }

    @GetMapping("/decline")
    public ResponseEntity<Status> decline(
            @RequestParam final String workflowId,
            @RequestParam(defaultValue = "") final String notes) {
        final EngagementFlow stub = client.newRpcStub(EngagementFlow.class, workflowId);
        return ResponseEntity.ok(client.invokeRPC(stub::decline, notes));
    }

    @GetMapping("/accept")
    public ResponseEntity<Status> accept(
            @RequestParam final String workflowId,
            @RequestParam(defaultValue = "") final String notes) {
        final EngagementFlow stub = client.newRpcStub(EngagementFlow.class, workflowId);
        return ResponseEntity.ok(client.invokeRPC(stub::accept, notes));
    }

    @GetMapping("/list")
    public ResponseEntity<SearchFlowsPage> list(@RequestParam final String query) {
        return ResponseEntity.ok(client.searchFlows(query, 100, ""));
    }
}
